## CMPINF 2100: Homework 05

Trey Randall

## Assigned: Tuesday of Week 05 at 11:00PM

## DUE: Tuesday of Week 06 at 11:59PM

## Overview

This assignment is more open ended than the previous ones. You must explore three data sets using the skills you have learned up to this point in the course. The Week 05 lecture material discusses the ESSENTIAL Pandas methods methods that you must ALWAYS use when BEGINNING to explore data. You must use those methods here in this assignment. You must also use SPLIT-APPLY-COMBINE to explore if summary statistics change across grouping variables. The last problem requires you to use the ESSENTIAL Pandas methods to begin exploring data associated with one of the final projects of your choosing.

**You must download the 3 data sets provided in the Canvas assignment page and save them to the appropriate directory on your computer.**

## Required tasks for Problem 01 through 03

For each of the 3 assigned data sets you must perform the following ESSENTIAL activities:

* Display the number of rows and columns in the dataset  
* Display the names of the columns and their associated data types  
* Display the number of unique values for each column  
* Display the number of MISSING values for each column  
* Display summary statistics appropriate for the data type for each column
* Count the number of rows associated with each unique value for:
    * Non-numeric columns  
    * Numeric columns with low number of unique values  

You must also execute SPLIT-APPLY-COMBINE to summarize the data based on grouping variables. Specifically, you must group by non-numeric columns and calculate the following summary statistics:
* Calculate the total number of rows for numeric columns  
* Calculate the number of non missing rows for numeric columns
* Calculate the mean and standard deviation of numeric columns
* Calculate the standard error on the mean of numeric columns
* Calculate the number of unique values for non-numeric columns

You should consider grouping by numeric columns that have low number of unique values.  

For each grouped DataFrame, you must answer the following:  
* Does the average of the numeric variables CHANGE across the groups?  

### Collaborators

Type the names of other students you worked with here.

## Problem 00

You will work with the NumPy and Pandas modules in this assignment.

Import NumPy and Pandas using their commonly accepted aliases.

#### 00) - SOLUTION

In [1]:
import numpy as np
import pandas as pd

In [2]:
import os

In [3]:
os.listdir()

['.ipynb_checkpoints',
 'homework_05_instructions.html',
 'homework_05_template.ipynb',
 'hw05_prob_01.csv',
 'hw05_prob_02.csv',
 'hw05_prob_03.csv',
 'Notes']

## Problem 01

Download `hw05_prob_01.csv` from the Canvas site. Save the CSV file in the same directory as this Jupyter notebook.

Read in the `hw05_prob_01.csv` CSV file and assign it to the `df01` object.

Explore the `df01` object following **ALL** required actions listed previously.

You must add more code cells and markdown cells to complete this problem.

#### 01) - SOLUTION

## Read in the CSV Data

In [7]:
pd.read_csv('hw05_prob_01.csv')

A   B   C         x
0    a1  b0  C1 -6.240387
1    a1  b0  C1 -4.860661
2    a1  b0  C1 -5.066712
3    a1  b0  C1 -5.635156
4    a1  b0  C1 -4.893876
..   ..  ..  ..       ...
125  a2  b1  C2  1.155377
126  a2  b1  C2  2.818466
127  a2  b1  C2  3.294759
128  a2  b1  C2  2.398190
129  a2  b1  C2  2.527589

[130 rows x 4 columns]

In [8]:
df01=pd.read_csv('hw05_prob_01.csv')

## Display number of rows and columns 

In [9]:
df01.shape

(130, 4)

In [11]:
df01.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   A       130 non-null    object 
 1   B       130 non-null    object 
 2   C       130 non-null    object 
 3   x       130 non-null    float64
dtypes: float64(1), object(3)
memory usage: 4.2+ KB


There are 130 rows and 4 columns. 

## Display Names of Columns and Data Types

In [40]:
df01.columns

Index(['A', 'B', 'C', 'x'], dtype='object')

In [41]:
df01.dtypes

A     object
B     object
C     object
x    float64
dtype: object

Columns A, B, and C are strings and column x is a Float

## Display the unique values for each column

In [42]:
df01.nunique()

A      2
B      2
C      2
x    130
dtype: int64

Columns A, B, and C each have two unique values, and column x has 130. 

## Display the number of Missing Vales for each Column

In [43]:
df01.isna()

A      B      C      x
0    False  False  False  False
1    False  False  False  False
2    False  False  False  False
3    False  False  False  False
4    False  False  False  False
..     ...    ...    ...    ...
125  False  False  False  False
126  False  False  False  False
127  False  False  False  False
128  False  False  False  False
129  False  False  False  False

[130 rows x 4 columns]

In [44]:
df01.isna().sum()

A    0
B    0
C    0
x    0
dtype: int64

There are no missing values in the dataset. 

## Display Summary Statistics appropriate for the data type for each column

In [45]:
df01.describe()

x
count  130.000000
mean     2.155995
std      6.015095
min     -7.185043
25%     -0.715064
50%      0.828413
75%      4.375242
max     15.382198

In [46]:
df01.describe(include='object')

A    B    C
count   130  130  130
unique    2    2    2
top      a1   b1   C2
freq     68   78   72

## Count the Number of rows associated with each unique value for: Non-Number columns and Numberic columns with low number of unique values

In [49]:
df01.dtypes.value_counts()

object     3
float64    1
Name: count, dtype: int64

In [50]:
df01.A.value_counts()

A
a1    68
a2    62
Name: count, dtype: int64

In [51]:
df01.A.value_counts(dropna=False, normalize=True)

A
a1    0.523077
a2    0.476923
Name: proportion, dtype: float64

For column A, the value counts for a1 is 68 or 52.3%, and for a2, is 62 or 46.7%

In [52]:
df01.B.value_counts()

B
b1    78
b0    52
Name: count, dtype: int64

In [53]:
df01.B.value_counts(dropna=False, normalize=True)

B
b1    0.6
b0    0.4
Name: proportion, dtype: float64

For column B, the value counts for b1 is 78 or 60%, and for b0, is 52 or 40%

In [54]:
df01.C.value_counts()

C
C2    72
C1    58
Name: count, dtype: int64

In [55]:
df01.C.value_counts(dropna=False, normalize=True)

C
C2    0.553846
C1    0.446154
Name: proportion, dtype: float64

For column C, the value counts for C1 is 58 or 44.6%, and for C2, is 72 or 55.4%

## SPLIT - APPLY - COMBINE

In [57]:
df01.groupby('A', dropna=False).\
aggregate(x_numrows = ('x', 'size'),
          x_nonmissing = ('x','count'),
          x_avg = ('x','mean'),
          x_std = ('x', 'std'),
          x_sem = ('x', 'sem'),
          B_nunique = ('B','nunique'),
          C_nunique = ('C','nunique')).\
reset_index()

A  x_numrows  x_nonmissing     x_avg     x_std    x_sem  B_nunique  \
0  a1         68            68  3.916666  6.963599  0.84446          2   
1  a2         62            62  0.224937  4.000938  0.50812          2   

   C_nunique  
0          2  
1          2

In [58]:
df01_A_summary_info =df01.groupby('A', dropna=False).\
aggregate(x_numrows = ('x', 'size'),
          x_nonmissing = ('x','count'),
          x_avg = ('x','mean'),
          x_std = ('x', 'std'),
          x_sem = ('x', 'sem'),
          B_nunique = ('B','nunique'),
          C_nunique = ('C','nunique')).\
reset_index()

In [59]:
df01_A_summary_info

A  x_numrows  x_nonmissing     x_avg     x_std    x_sem  B_nunique  \
0  a1         68            68  3.916666  6.963599  0.84446          2   
1  a2         62            62  0.224937  4.000938  0.50812          2   

   C_nunique  
0          2  
1          2

Yes, the average of 'x' does change across the groups of column A. The difference here is significant and would imply that there are different characteristics across the group.   

## Problem 02

Download `hw05_prob_02.csv` from the Canvas site. Save the CSV file in the same directory as this Jupyter notebook.

Read in the `hw05_prob_02.csv` CSV file and assign it to the `df02` object.

Explore the `df02` object following **ALL** required actions listed previously.

You must add more code cells and markdown cells to complete this problem.

#### 02) - SOLUTION

## Read in the CSV data

In [60]:
pd.read_csv('hw05_prob_02.csv')

x         y    A   G
0    0.249430  1.226548  2.0   1
1   -1.049282 -1.100737  NaN   1
2   -0.147525 -0.679438  NaN   1
3    1.053127 -0.475298  NaN   1
4    1.242744 -1.060518  2.0   1
..        ...       ...  ...  ..
363  0.357651 -0.608975  NaN  14
364  0.235176  0.415023  NaN  14
365  0.283675 -0.981750  2.0  14
366  0.929657 -0.546980  2.0  14
367 -1.068429  1.795239  2.0  14

[368 rows x 4 columns]

In [63]:
df02=pd.read_csv('hw05_prob_02.csv')

## Display number of rows and columns 

In [64]:
df02.shape

(368, 4)

In [65]:
df02.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 368 entries, 0 to 367
Data columns (total 4 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   x       368 non-null    float64
 1   y       352 non-null    float64
 2   A       347 non-null    float64
 3   G       368 non-null    int64  
dtypes: float64(3), int64(1)
memory usage: 11.6 KB


There are 368 rows and 4 columns in df02. 

## Display Names of Columns and Data Types

In [66]:
df02.columns

Index(['x', 'y', 'A', 'G'], dtype='object')

In [67]:
df02.dtypes

x    float64
y    float64
A    float64
G      int64
dtype: object

Columns x, y, and A are floats and column G is an integer.

## Display the unique values for each column

In [68]:
df02.nunique()

x    368
y    352
A      2
G     14
dtype: int64

Column x has 368 unique values, column y has 352, column A has 2, and column G has 14. 

## Display the number of Missing Vales for each Column

In [69]:
df02.isna()

x      y      A      G
0    False  False  False  False
1    False  False   True  False
2    False  False   True  False
3    False  False   True  False
4    False  False  False  False
..     ...    ...    ...    ...
363  False  False   True  False
364  False  False   True  False
365  False  False  False  False
366  False  False  False  False
367  False  False  False  False

[368 rows x 4 columns]

In [71]:
df02.isna().sum()

x     0
y    16
A    21
G     0
dtype: int64

Column x and G have no missing values, whereas column y has 16 and A has 21. 

## Display Summary Statistics appropriate for the data type for each column

In [72]:
df02.describe()

x           y           A           G
count  368.000000  352.000000  347.000000  368.000000
mean     0.008874    0.287405    1.608069    6.744565
std      1.017163    1.951770    0.488886    4.121450
min     -2.698471   -7.737557    1.000000    1.000000
25%     -0.696962   -0.599940    1.000000    3.000000
50%     -0.027857    0.188779    2.000000    6.000000
75%      0.607358    1.331244    2.000000   10.000000
max      2.936346    7.875071    2.000000   14.000000

## Count the Number of rows associated with each unique value for: Non-Number columns and Numberic columns with low number of unique values

In [74]:
df02.dtypes.value_counts()

float64    3
int64      1
Name: count, dtype: int64

In [78]:
df02.A.value_counts(dropna=False)

A
2.0    211
1.0    136
NaN     21
Name: count, dtype: int64

In [79]:
df02.A.value_counts(dropna=False, normalize=True)

A
2.0    0.573370
1.0    0.369565
NaN    0.057065
Name: proportion, dtype: float64

For column A, the value counts for 1.0 is 136 or 37%; for a2, it is 211 or 57.3%, and finally, there are 21 missing values, which account for 5.7%.

In [80]:
df02.G.value_counts(dropna=False)

G
5     45
10    45
2     39
1     38
3     35
14    30
11    29
7     25
9     19
4     15
6     15
8     14
13    11
12     8
Name: count, dtype: int64

In [81]:
df02.G.value_counts(dropna=False, normalize = True)

G
5     0.122283
10    0.122283
2     0.105978
1     0.103261
3     0.095109
14    0.081522
11    0.078804
7     0.067935
9     0.051630
4     0.040761
6     0.040761
8     0.038043
13    0.029891
12    0.021739
Name: proportion, dtype: float64

## SPLIT - APPLY - COMBINE

In [93]:
df02.groupby('A', dropna=False).\
aggregate(x_numrows = ('x', 'size'),
          x_nonmissing = ('x','count'),
          x_avg = ('x','mean'),
          x_std = ('x', 'std'),
          x_sem = ('x', 'sem'),
          y_numrows = ('y', 'size'),
          y_nonmissing = ('y','count'),
          y_avg = ('y','mean'),
          y_std = ('y', 'std'),
          y_sem = ('y', 'sem'),
          G_numrows = ('G', 'size'),
          G_nonmissing = ('G','count'),
          G_avg = ('G','mean'),
          G_std = ('G', 'std'),
          G_sem = ('G', 'sem')).\
reset_index()

A  x_numrows  x_nonmissing     x_avg     x_std     x_sem  y_numrows  \
0  1.0        136           136 -0.071118  1.007348  0.086379        136   
1  2.0        211           211  0.029215  1.020511  0.070255        211   
2  NaN         21            21  0.322540  1.024403  0.223543         21   

   y_nonmissing     y_avg     y_std     y_sem  G_numrows  G_nonmissing  \
0           120  0.097019  1.935619  0.176697        136           136   
1           211  0.435802  2.025692  0.139454        211           211   
2            21 -0.115713  0.925330  0.201923         21            21   

       G_avg     G_std     G_sem  
0   8.573529  2.671197  0.229053  
1   5.028436  3.871649  0.266535  
2  12.142857  4.661392  1.017199

In [94]:
df02_A_summary_info = df02.groupby('A', dropna=False).\
aggregate(x_numrows = ('x', 'size'),
          x_nonmissing = ('x','count'),
          x_avg = ('x','mean'),
          x_std = ('x', 'std'),
          x_sem = ('x', 'sem'),
          y_numrows = ('y', 'size'),
          y_nonmissing = ('y','count'),
          y_avg = ('y','mean'),
          y_std = ('y', 'std'),
          y_sem = ('y', 'sem'),
          G_numrows = ('G', 'size'),
          G_nonmissing = ('G','count'),
          G_avg = ('G','mean'),
          G_std = ('G', 'std'),
          G_sem = ('G', 'sem')).\
reset_index()

In [95]:
df02_A_summary_info

A  x_numrows  x_nonmissing     x_avg     x_std     x_sem  y_numrows  \
0  1.0        136           136 -0.071118  1.007348  0.086379        136   
1  2.0        211           211  0.029215  1.020511  0.070255        211   
2  NaN         21            21  0.322540  1.024403  0.223543         21   

   y_nonmissing     y_avg     y_std     y_sem  G_numrows  G_nonmissing  \
0           120  0.097019  1.935619  0.176697        136           136   
1           211  0.435802  2.025692  0.139454        211           211   
2            21 -0.115713  0.925330  0.201923         21            21   

       G_avg     G_std     G_sem  
0   8.573529  2.671197  0.229053  
1   5.028436  3.871649  0.266535  
2  12.142857  4.661392  1.017199

Yes, the averages for 'x', 'y', and 'G' all change across the groups of column A. For both 'x' and 'y', the average variance across the values in column A shows a small variance, and for columns 'y' and 'G', all the values are positive, which could indicate that the the values in these groups tend to be positive. 

## Problem 03

Download `hw05_prob_03.csv` from the Canvas site. Save the CSV file in the same directory as this Jupyter notebook.

Read in the `hw05_prob_03.csv` CSV file and assign it to the `df03` object.

Explore the `df03` object following **ALL** required actions listed previously.

You must add more code cells and markdown cells to complete this problem.

#### 03) - SOLUTION

## Read in the CSV data

In [96]:
pd.read_csv('hw05_prob_03.csv')

obs_id key     value
0          1  x1  0.034632
1          2  x1  0.523494
2          3  x1  1.224766
3          4  x1  1.517422
4          5  x1  2.432589
...      ...  ..       ...
1402     197  x7 -8.145633
1403     198  x7       NaN
1404     199  x7 -8.152787
1405     200  x7 -8.063927
1406     201  x7       NaN

[1407 rows x 3 columns]

In [97]:
df03=pd.read_csv('hw05_prob_03.csv')

## Display number of rows and columns

In [98]:
df03.shape

(1407, 3)

In [99]:
df03.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1407 entries, 0 to 1406
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   obs_id  1407 non-null   int64  
 1   key     1407 non-null   object 
 2   value   1379 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 33.1+ KB


There are 1407 rows and 3 columns in df03. 

## Display Names of Columns and Data Types

In [100]:
df03.columns

Index(['obs_id', 'key', 'value'], dtype='object')

In [101]:
df03.dtypes

obs_id      int64
key        object
value     float64
dtype: object

Column obs_id is an integer, the column value is a float, and a column key is an object.

## Display the unique values for each column

In [102]:
df03.nunique()

obs_id     201
key          7
value     1379
dtype: int64

Column obs_id has 201 unique values, column key has 7, and column values have 1379. 

## Display the number of Missing Vales for each Column

In [103]:
df03.isna()

obs_id    key  value
0      False  False  False
1      False  False  False
2      False  False  False
3      False  False  False
4      False  False  False
...      ...    ...    ...
1402   False  False  False
1403   False  False   True
1404   False  False  False
1405   False  False  False
1406   False  False   True

[1407 rows x 3 columns]

In [104]:
df03.isna().sum()

obs_id     0
key        0
value     28
dtype: int64

Columns obs_id and key have no missing values, whereas column value has 28.

## Display Summary Statistics appropriate for the data type for each column

In [105]:
df03.describe()

obs_id        value
count  1407.000000  1379.000000
mean    101.000000    -2.222709
std      58.043614     4.465882
min       1.000000   -25.362015
25%      51.000000    -2.683234
50%     101.000000    -0.798222
75%     151.000000     0.250730
max     201.000000     4.209716

In [106]:
df03.describe(include='object')

key
count   1407
unique     7
top       x1
freq     201

## Count the Number of rows associated with each unique value for: Non-Number columns and Numberic columns with low number of unique values

In [107]:
df03.dtypes.value_counts()

int64      1
object     1
float64    1
Name: count, dtype: int64

In [109]:
df03.key.value_counts(dropna=False)

key
x1    201
x2    201
x3    201
x4    201
x5    201
x6    201
x7    201
Name: count, dtype: int64

In [112]:
df03.key.value_counts(dropna=False, normalize=True)

key
x1    0.142857
x2    0.142857
x3    0.142857
x4    0.142857
x5    0.142857
x6    0.142857
x7    0.142857
Name: proportion, dtype: float64

For the unique values in key x1 - x7, each one appears 201 times, which gives a proportion of 14.3%

## SPLIT - APPLY - COMBINE

In [114]:
df03.groupby('key', dropna=False).\
aggregate(obs_id_numrows = ('obs_id', 'size'),
          obs_id_nonmissing = ('obs_id','count'),
          obs_id_avg = ('obs_id','mean'),
          obs_id_std = ('obs_id', 'std'),
          obs_id_sem = ('obs_id', 'sem'),
          value_numrows = ('value', 'size'),
          value_nonmissing = ('value','count'),
          value_avg = ('value','mean'),
          value_std = ('value', 'std'),
          value_sem = ('value', 'sem')).\
reset_index()

key  obs_id_numrows  obs_id_nonmissing  obs_id_avg  obs_id_std  obs_id_sem  \
0  x1             201                201       101.0   58.167861    4.102845   
1  x2             201                201       101.0   58.167861    4.102845   
2  x3             201                201       101.0   58.167861    4.102845   
3  x4             201                201       101.0   58.167861    4.102845   
4  x5             201                201       101.0   58.167861    4.102845   
5  x6             201                201       101.0   58.167861    4.102845   
6  x7             201                201       101.0   58.167861    4.102845   

   value_numrows  value_nonmissing  value_avg  value_std  value_sem  
0            201               201  -0.160787   1.735903   0.122441  
1            201               201   0.100830   0.095558   0.006740  
2            201               197  -0.050132   1.544633   0.110051  
3            201               201 -10.187219   5.079706   0.358295  
4            201               201   1.061053   0.935690   0.065998  
5            201               197  -1.463831   0.591750   0.042160  
6            201               181  -5.085382   3.002404   0.223167

In [115]:
df03_key_summary_info = df03.groupby('key', dropna=False).\
aggregate(obs_id_numrows = ('obs_id', 'size'),
          obs_id_nonmissing = ('obs_id','count'),
          obs_id_avg = ('obs_id','mean'),
          obs_id_std = ('obs_id', 'std'),
          obs_id_sem = ('obs_id', 'sem'),
          value_numrows = ('value', 'size'),
          value_nonmissing = ('value','count'),
          value_avg = ('value','mean'),
          value_std = ('value', 'std'),
          value_sem = ('value', 'sem')).\
reset_index()

In [116]:
df03_key_summary_info

key  obs_id_numrows  obs_id_nonmissing  obs_id_avg  obs_id_std  obs_id_sem  \
0  x1             201                201       101.0   58.167861    4.102845   
1  x2             201                201       101.0   58.167861    4.102845   
2  x3             201                201       101.0   58.167861    4.102845   
3  x4             201                201       101.0   58.167861    4.102845   
4  x5             201                201       101.0   58.167861    4.102845   
5  x6             201                201       101.0   58.167861    4.102845   
6  x7             201                201       101.0   58.167861    4.102845   

   value_numrows  value_nonmissing  value_avg  value_std  value_sem  
0            201               201  -0.160787   1.735903   0.122441  
1            201               201   0.100830   0.095558   0.006740  
2            201               197  -0.050132   1.544633   0.110051  
3            201               201 -10.187219   5.079706   0.358295  
4            201               201   1.061053   0.935690   0.065998  
5            201               197  -1.463831   0.591750   0.042160  
6            201               181  -5.085382   3.002404   0.223167

The average for the column obs_id does not change, indicating that this number is a constant unique value. This is proven by looking at the unique value method. The average for the value column does change, and some numbers change dramatically. For key x4, there could be some data that could be considered an outlier. For x2 and x5, there is a positive average, which could indicate more positive values, whereas the other columns would have more negative values. 

## Problem 04

You must download the data associated with one of the Final Projects from the Canvas site. Save the file(s) in the same directory as this Jupyter notebook.

Read in the data associated with one of the Final Projects. You must perform the ESSENTIAL activities listed previously.

You do **NOT** need to SPLIT-APPLY-COMBINE to explore across groups for this assignment.

#### 04) - SOLUTION

## Read in the CSV data - Project Name: Pitt: Calibrating Additive Manufacturing (AM) Simulations

In [117]:
pd.read_csv('training_inputs.csv')

sim_id           x1        x2         x3         x4
0        1  1428.262963  0.010276  76.449325  22.963241
1        2  1119.096654  0.001542  49.076221   6.858423
2        3   910.521261  0.005973  57.298239  25.742151
3        4  1258.042700  0.008673  79.834414  14.377447
4        5  1068.591510  0.015057  33.380801  21.477560
5        6  1321.991511  0.001337  65.498750  18.325935
6        7  1226.342504  0.019976  56.402869   7.789828
7        8   983.086199  0.016044  53.258946   2.530931
8        9  1387.548586  0.005605  63.950159  28.052451
9       10  1294.835000  0.007688  30.451657  19.912775
10      11   848.652481  0.009394  38.233815  21.142945
11      12  1491.847697  0.004059  44.484207  24.221998
12      13  1093.418310  0.004134  75.424970  23.546394
13      14  1421.824769  0.013740  40.535213  27.462553
14      15  1525.011126  0.012388  56.214007  26.182474
15      16  1124.177780  0.002455  71.293726   8.253403
16      17   892.238232  0.003249  66.700567  15.814664
17      18   864.569145  0.010943  71.968422  20.440992
18      19   878.954905  0.017174  42.621494  13.366830
19      20   807.788620  0.012753  58.572954  10.234053
20      21  1481.916293  0.004990  47.608519   3.446863
21      22  1182.003808  0.002853  50.922495  27.152865
22      23  1557.147105  0.009790  37.787333  14.861847
23      24  1343.090092  0.006406  32.176282   7.623467
24      25  1236.531345  0.012506  34.848645   2.761267
25      26  1403.288931  0.016408  33.909144  17.351408
26      27  1585.754835  0.006895  59.977589  17.525933
27      28  1011.777710  0.013379  50.235776  29.522479
28      29  1371.032588  0.001797  46.101859  14.045899
29      30  1162.660181  0.018380  36.992893  10.751260
30      31  1145.651429  0.011768  69.327816  29.166249
31      32  1337.828389  0.014311  60.326921   0.330430
32      33  1533.946513  0.017786  52.255697  16.109363
33      34  1501.715605  0.003569  68.232295   9.762321
34      35  1550.955238  0.014097  70.792394  11.607739
35      36   818.574666  0.014783  54.749260  22.379004
36      37  1573.296535  0.011046  54.194362   6.063359
37      38   841.630716  0.004536  49.804795  11.282154
38      39  1085.579854  0.007505  45.161910   1.034904
39      40   953.696301  0.009270  77.689741   8.993805
40      41  1287.526938  0.018156  74.437610  12.470051
41      42  1275.806517  0.005423  63.448286   1.385522
42      43  1442.232585  0.016597  42.799373   5.159088
43      44  1218.899297  0.019236  46.601845  24.984234
44      45  1364.163428  0.010519  75.696766   4.084601
45      46  1205.539408  0.018745  64.977447  25.317988
46      47  1057.338678  0.015431  78.414147  18.805273
47      48  1171.201196  0.008441  39.347081  28.735387
48      49   931.782344  0.008218  62.509521   2.181161
49      50  1456.945272  0.017430  67.331467  21.996942
50      51  1044.531147  0.015836  72.774337   4.715959
51      52   964.972365  0.019312  61.264318  16.758174
52      53  1024.257529  0.006707  31.206395  12.791722
53      54   924.147836  0.011490  35.952168   5.957354
54      55  1003.148766  0.002233  41.698575  19.541825

In [118]:
df_training_inputs = pd.read_csv('training_inputs.csv')

In [119]:
pd.read_csv('training_outputs.csv')

sim_id  layer_id  action_id        time          y
0          1         1          1    0.010276  1428.2600
1          1         1          2    9.310280    86.1325
2          1         1          3   12.946600    76.1949
3          1         2          1   12.956900  1428.2600
4          1         2          2   22.256900   116.2790
...      ...       ...        ...         ...        ...
3295       9        19          2  242.261000   157.5390
3296       9        19          3  245.897000   146.9260
3297       9        20          1  245.903000  1387.5500
3298       9        20          2  255.203000   158.9100
3299       9        20          3  258.839000   148.3480

[3300 rows x 5 columns]

In [120]:
df_training_outputs = pd.read_csv('training_outputs.csv')

## Display number of rows and columns - Training Input Data

In [121]:
df_training_inputs.shape

(55, 5)

In [122]:
df_training_inputs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 55 entries, 0 to 54
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   sim_id  55 non-null     int64  
 1   x1      55 non-null     float64
 2   x2      55 non-null     float64
 3   x3      55 non-null     float64
 4   x4      55 non-null     float64
dtypes: float64(4), int64(1)
memory usage: 2.3 KB


There are 55 rows and 5 columns in Training Input Data.

## Display number of rows and columns - Training Output Data

In [123]:
df_training_outputs.shape

(3300, 5)

In [124]:
df_training_outputs.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3300 entries, 0 to 3299
Data columns (total 5 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   sim_id     3300 non-null   int64  
 1   layer_id   3300 non-null   int64  
 2   action_id  3300 non-null   int64  
 3   time       3300 non-null   float64
 4   y          3300 non-null   float64
dtypes: float64(2), int64(3)
memory usage: 129.0 KB


There are 3300 rows and 5 columns in Training Output Data.

## Display Names of Columns and Data Types - Training Input Data

In [125]:
df_training_inputs.columns

Index(['sim_id', 'x1', 'x2', 'x3', 'x4'], dtype='object')

In [130]:
df_training_inputs.dtypes

sim_id      int64
x1        float64
x2        float64
x3        float64
x4        float64
dtype: object

Column sim_id is an integer. The other columns (x1, x2, x3, x4) are all floats.

## Display Names of Columns and Data Types - Training Output Data

In [131]:
df_training_outputs.columns

Index(['sim_id', 'layer_id', 'action_id', 'time', 'y'], dtype='object')

In [132]:
df_training_outputs.dtypes

sim_id         int64
layer_id       int64
action_id      int64
time         float64
y            float64
dtype: object

Columns sim_id, layer_id, and action_id are all integers. Columns time and y are floats. 

## Display the unique values for each column - Training Input Data

In [133]:
df_training_inputs.nunique()

sim_id    55
x1        55
x2        55
x3        55
x4        55
dtype: int64

All of the inputs have 55 unique values.

## Display the unique values for each column - Training Output Data

In [134]:
df_training_outputs.nunique()

sim_id         55
layer_id       20
action_id       3
time         3143
y            2243
dtype: int64

Sim_id has 55 unique values, layer_id has 20, and action_id has 3. The time and y columns have more, with 3143 and 2243, respectively. 

## Display the number of Missing Vales for each Column - Training Input Data

In [135]:
df_training_inputs.isna()

sim_id     x1     x2     x3     x4
0    False  False  False  False  False
1    False  False  False  False  False
2    False  False  False  False  False
3    False  False  False  False  False
4    False  False  False  False  False
5    False  False  False  False  False
6    False  False  False  False  False
7    False  False  False  False  False
8    False  False  False  False  False
9    False  False  False  False  False
10   False  False  False  False  False
11   False  False  False  False  False
12   False  False  False  False  False
13   False  False  False  False  False
14   False  False  False  False  False
15   False  False  False  False  False
16   False  False  False  False  False
17   False  False  False  False  False
18   False  False  False  False  False
19   False  False  False  False  False
20   False  False  False  False  False
21   False  False  False  False  False
22   False  False  False  False  False
23   False  False  False  False  False
24   False  False  False  False  False
25   False  False  False  False  False
26   False  False  False  False  False
27   False  False  False  False  False
28   False  False  False  False  False
29   False  False  False  False  False
30   False  False  False  False  False
31   False  False  False  False  False
32   False  False  False  False  False
33   False  False  False  False  False
34   False  False  False  False  False
35   False  False  False  False  False
36   False  False  False  False  False
37   False  False  False  False  False
38   False  False  False  False  False
39   False  False  False  False  False
40   False  False  False  False  False
41   False  False  False  False  False
42   False  False  False  False  False
43   False  False  False  False  False
44   False  False  False  False  False
45   False  False  False  False  False
46   False  False  False  False  False
47   False  False  False  False  False
48   False  False  False  False  False
49   False  False  False  False  False
50   False  False  False  False  False
51   False  False  False  False  False
52   False  False  False  False  False
53   False  False  False  False  False
54   False  False  False  False  False

In [136]:
df_training_inputs.isna().sum()

sim_id    0
x1        0
x2        0
x3        0
x4        0
dtype: int64

The training input data has no missing values.

## Display the number of Missing Vales for each Column - Training Output Data

In [138]:
df_training_outputs.isna()

sim_id  layer_id  action_id   time      y
0      False     False      False  False  False
1      False     False      False  False  False
2      False     False      False  False  False
3      False     False      False  False  False
4      False     False      False  False  False
...      ...       ...        ...    ...    ...
3295   False     False      False  False  False
3296   False     False      False  False  False
3297   False     False      False  False  False
3298   False     False      False  False  False
3299   False     False      False  False  False

[3300 rows x 5 columns]

In [139]:
df_training_outputs.isna().sum()

sim_id       0
layer_id     0
action_id    0
time         0
y            0
dtype: int64

The training outputs data also has no missing values.

## Display Summary Statistics appropriate for the data type for each column - Training Input Data

In [140]:
df_training_inputs.describe()

sim_id           x1         x2         x3         x4
count  55.00000    55.000000  55.000000  55.000000  55.000000
mean   28.00000  1199.952243   0.010515  54.967379  15.006454
std    16.02082   232.472509   0.005548  14.537852   8.699401
min     1.00000   807.788620   0.001337  30.451657   0.330430
25%    14.50000  1007.463238   0.005789  42.710434   7.706647
50%    28.00000  1205.539408   0.010519  54.749260  14.861847
75%    41.50000  1395.418758   0.015244  67.016017  22.187973
max    55.00000  1585.754835   0.019976  79.834414  29.522479

## Display Summary Statistics appropriate for the data type for each column - Training Output Data

In [141]:
df_training_outputs.describe()

sim_id     layer_id   action_id         time            y
count  3300.000000  3300.000000  3300.00000  3300.000000  3300.000000
mean     28.000000    10.500000     2.00000   130.417977   490.807334
std      15.876914     5.767155     0.81662    74.865191   519.994476
min       1.000000     1.000000     1.00000     0.001337    48.557500
25%      14.000000     5.750000     1.00000    64.739025   117.711500
50%      28.000000    10.500000     2.00000   129.473500   165.163500
75%      42.000000    15.250000     3.00000   194.210000  1003.150000
max      55.000000    20.000000     3.00000   259.127000  1585.750000

## Count the Number of rows associated with each unique value for: Non-Number columns and Numberic columns with low number of unique values - Training Input Data

In [142]:
df_training_inputs.dtypes.value_counts()

float64    4
int64      1
Name: count, dtype: int64

In [143]:
df_training_inputs.sim_id.value_counts(dropna=False)

sim_id
1     1
42    1
31    1
32    1
33    1
34    1
35    1
36    1
37    1
38    1
39    1
40    1
41    1
43    1
29    1
44    1
45    1
46    1
47    1
48    1
49    1
50    1
51    1
52    1
53    1
54    1
30    1
28    1
2     1
14    1
3     1
4     1
5     1
6     1
7     1
8     1
9     1
10    1
11    1
12    1
13    1
15    1
27    1
16    1
17    1
18    1
19    1
20    1
21    1
22    1
23    1
24    1
25    1
26    1
55    1
Name: count, dtype: int64

In [144]:
df_training_inputs.sim_id.value_counts(dropna=False, normalize=True)

sim_id
1     0.018182
42    0.018182
31    0.018182
32    0.018182
33    0.018182
34    0.018182
35    0.018182
36    0.018182
37    0.018182
38    0.018182
39    0.018182
40    0.018182
41    0.018182
43    0.018182
29    0.018182
44    0.018182
45    0.018182
46    0.018182
47    0.018182
48    0.018182
49    0.018182
50    0.018182
51    0.018182
52    0.018182
53    0.018182
54    0.018182
30    0.018182
28    0.018182
2     0.018182
14    0.018182
3     0.018182
4     0.018182
5     0.018182
6     0.018182
7     0.018182
8     0.018182
9     0.018182
10    0.018182
11    0.018182
12    0.018182
13    0.018182
15    0.018182
27    0.018182
16    0.018182
17    0.018182
18    0.018182
19    0.018182
20    0.018182
21    0.018182
22    0.018182
23    0.018182
24    0.018182
25    0.018182
26    0.018182
55    0.018182
Name: proportion, dtype: float64

## Count the Number of rows associated with each unique value for: Non-Number columns and Numberic columns with low number of unique values - Training Output Data

In [145]:
df_training_outputs.dtypes.value_counts()

int64      3
float64    2
Name: count, dtype: int64

In [146]:
df_training_outputs.action_id.value_counts(dropna=False)

action_id
1    1100
2    1100
3    1100
Name: count, dtype: int64

In [147]:
df_training_outputs.action_id.value_counts(dropna=False, normalize=True)

action_id
1    0.333333
2    0.333333
3    0.333333
Name: proportion, dtype: float64